### PR TITLE
TST: Check error raised when inserting wrong length categorical column

### LINKED ISSUE #9374: BUG: inserting a Categorical with the wrong length into a DataFrame is allowed
This leaves the DataFrame in a very weird/buggy state:

```
In [33]: cat = pd.Categorical.from_codes([0, 1, 1, 0, 1, 2], ['a', 'b', 'c'])

In [34]: df = pd.DataFrame()

In [35]: df['bar'] = range(10)

In [36]: df['foo'] = cat

In [37]: df
Out[37]:
   bar foo
0    0   a
1    1   b
2    2   b
3    3   a
4    4   b
5    5   c
6    6
7    7
8    8
9    9

In [38]: df.foo.shape
Out[38]: (6,)
```

I was expecting something like:

```
In [49]: df['foo'] = np.array(cat)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-49-2c5dea325c23> in <module>()
----> 1 df['foo'] = np.array(cat)

/Users/shoyer/dev/pandas/pandas/core/frame.py in __setitem__(self, key, value)
   2108         else:
   2109             # set column
-> 2110             self._set_item(key, value)
   2111
   2112     def _setitem_slice(self, key, value):

/Users/shoyer/dev/pandas/pandas/core/frame.py in _set_item(self, key, value)
   2185
   2186         self._ensure_valid_index(value)
-> 2187         value = self._sanitize_column(key, value)
   2188         NDFrame._set_item(self, key, value)
   2189

/Users/shoyer/dev/pandas/pandas/core/frame.py in _sanitize_column(self, key, value)
   2258         elif (isinstance(value, Index) or is_sequence(value)):
   2259             from pandas.core.series import _sanitize_index
-> 2260             value = _sanitize_index(value, self.index, copy=False)
   2261             if not isinstance(value, (np.ndarray, Index)):
   2262                 if isinstance(value, list) and len(value) > 0:

/Users/shoyer/dev/pandas/pandas/core/series.py in _sanitize_index(data, index, copy)
   2562
   2563     if len(data) != len(index):
-> 2564         raise ValueError('Length of values does not match length of '
   2565                          'index')
   2566

ValueError: Length of values does not match length of index
```

Tested on master.

### PR DESCRIPTION
**Pull request by ryankarlos**
- [x] closes #9374
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

File: pandas/tests/frame/test_dtypes.py [hunk 1]
```diff
@@ -724,6 +724,13 @@ def test_astype_duplicate_col(self):
         expected = concat([a1_str, b, a2_str], axis=1)
         tm.assert_frame_equal(result, expected)
 
+    def test_wrong_length_cat_dtype_raises(self):
+        cat = pd.Categorical.from_codes([0, 1, 1, 0, 1, 2], ["a", "b", "c"])
```
Comment (mroeschke) [inline pandas/tests/frame/test_dtypes.py]: Could you add the issue number here?

Comment (ryankarlos) [reply pandas/tests/frame/test_dtypes.py]: Done

**Comment by ryankarlos** [comment]
Any ides why CI is failing on Linux Py36 ?  Doesn't seem related to my additions

```
=================================== FAILURES ===================================
______________ test_chunks_have_consistent_numerical_type[python] ______________
[gw0] linux -- Python 3.6.1 /home/vsts/miniconda3/envs/pandas-dev/bin/python

all_parsers = <pandas.tests.io.parser.conftest.PythonParser object at 0x7fb4c86cb5f8>

    def test_chunks_have_consistent_numerical_type(all_parsers):
        parser = all_parsers
        integers = [str(i) for i in range(499999)]
        data = "a\n" + "\n".join(integers + ["1.0", "2.0"] + integers)
    
        # Coercions should work without warnings.
        with tm.assert_produces_warning(None):
>           result = parser.read_csv(StringIO(data))

pandas/tests/io/parser/test_common.py:1204: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <contextlib._GeneratorContextManager object at 0x7fb48ad70390>
type = None, value = None, traceback = None

    def __exit__(self, type, value, traceback):
        if type is None:
            try:
>               next(self.gen)
E               AssertionError: Caused unexpected warning(s): [('ResourceWarning', ResourceWarning("unclosed file <_io.BufferedReader name='test1.xlsx'>",), '/home/vsts/work/1/s/pandas/io/parsers.py', 2927)].

```

**Comment by mroeschke** [comment]
Just a flakey test: https://github.com/pandas-dev/pandas/issues/29514

Could you try pushing again?

**Comment by ryankarlos** [comment]
Ok just repushed - waiting for CI

**Comment by alimcmaster1** [comment]
CI failures still related to that test #29514

**Comment by ryankarlos** [comment]
Thanks - will try repushing again and see if it works  now.

File: pandas/tests/frame/test_dtypes.py [hunk 2]
```diff
@@ -724,6 +724,14 @@ def test_astype_duplicate_col(self):
         expected = concat([a1_str, b, a2_str], axis=1)
         tm.assert_frame_equal(result, expected)
 
+    def test_wrong_length_cat_dtype_raises(self):
+        # GH29523
+        cat = pd.Categorical.from_codes([0, 1, 1, 0, 1, 2], ["a", "b", "c"])
+        df = pd.DataFrame({"bar": range(10)})
+        err = "Length of values does not match length of index"
```
Comment (jreback) [inline pandas/tests/frame/test_dtypes.py]: can you move this test to pandas/tests/frame/test_indexing.py in the class TestDataFrameIndexingCategorical

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/frame/test_dtypes.py (hunk 2 above)
Comment (ryankarlos) [reply pandas/tests/frame/test_dtypes.py]: Done

**Review by mroeschke — APPROVED** [review]

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @ryankarlos

**Pull request merged**